src/pcap/pcap.go

```diff
@@ -0,0 +1,115 @@
+// Package pcap provides types and interfaces to build a remote packet capturing
+// tool. For details about the different applications see the different packages
+// in `cmd/`.
+package pcap
+
+import (
+	"fmt"
+	"unicode"
+
+	"github.com/go-playground/validator/v10"
+	"go.uber.org/zap"
+)
+
+//go:generate protoc --go_out=. --go_opt=paths=source_relative --go-grpc_out=. --go-grpc_opt=paths=source_relative pcap.proto
+
+const (
+	// CompatibilityLevel indicates whether two parties are compatible. Once there is a change
+	// that requires both parties to be updated this value MUST be incremented by one. The calling
+	// party has to ensure that the compatibility level of the called party is equal or larger and
+	// refuse operation if it isn't.
+	CompatibilityLevel int64 = 0
+
+	// LogKeyVcapID sets on which field the vcap request id will be logged.
+	LogKeyVcapID = "vcap-id"
+	HeaderVcapID = "x-vcap-request-id"
+)
+
+var (
+	errNilField       = fmt.Errorf("field is nil")
+	errInvalidPayload = fmt.Errorf("invalid payload")
+)
+
+// BufferConf allows to specify the behaviour of buffers.
+//
+// The recommendation is to set the upper limit slightly below the size
+// to account for data put into the buffer while checking the fill condition
+// or performing work. The lower limit should be low enough to make some room
+// for new data but not too low (which would cause a lot of data to be
+// discarded). After all the buffer should mainly soften short spikes in data
+// transfer and these limits only protect against uncontrolled back pressure.
+type BufferConf struct {
+	// Size is the number of responses that can be buffered per stream.
+	Size int `yaml:"size" validate:"gte=0"`
+	// UpperLimit tells the manager of the buffer to start discarding messages
+	// once the limit is exceeded. The condition looks like this:
+	//   len(buf) >= UpperLimit
+	UpperLimit int `yaml:"upperLimit" validate:"gte=0,ltefield=Size"`
```

Comment:
These are currently absolute values. Would it make sense to make those percentages if `Size`? Otherwise you need to adjust distinct numbers for 3 values, which would be in 80% of cases be derived from "95% of the buffer size" anyway.

Reply:
Initially I had it like that but you can get into weird rounding issues for small sizes and for large buffers you might not be able to leave a very tiny margin between upper limit and size. In this case I think it's better to have full control (also the code path that handles this logic looks way simpler this way).